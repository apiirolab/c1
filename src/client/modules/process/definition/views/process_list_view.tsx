import * as React from 'react';

import { Header, List, ListProps, Message } from 'semantic-ui-react';
import styled, { StyledComponentClass } from 'styled-components';

import { QueryTypes } from 'data/client';
import { ProcessItemView } from './process_item_view';

const ProcessList: StyledComponentClass<ListProps, {}> = styled(List)`
  max-width: 800px !important;

  &&&&& .description {
    font-size: smaller;
  }
`;

const ResultInfo = styled.div`
  color: #999;
`;

type Props = {
  processes: QueryTypes.BpmnProcessItem[];
  context: App.Context;
};

export const ProcessListView: React.SFC<Props> = ({ processes, context }) => (
  <React.Fragment>
    <ResultInfo>
      {processes.length === 1
        ? context.i18n`1 process available`
        : context.i18n`${processes.length.toString()} processes available`}
    </ResultInfo>
    <Header
      icon="tasks"
      content="Company Processes"
      subheader="Find and execute a new company process"
    />
    <Choose>
      <When condition={processes && processes.length > 0}>
        <ProcessList divided relaxed>
          {processes.map(process => (
            <ProcessItemView key={process.id} process={process} />
          ))}
        </ProcessList>
      </When>
      <Otherwise>
        <Message>No results</Message>
      </Otherwise>
    </Choose>
  </React.Fragment>
);

ProcessListView.displayName = 'ProcessListView';
