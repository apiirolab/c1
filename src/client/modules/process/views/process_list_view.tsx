import * as React from 'react';

import { Header, List, ListProps, Message } from 'semantic-ui-react';
import styled, { StyledComponentClass } from 'styled-components';

import { ProcessListItemView } from './process_list_item_view';

const VersionList: StyledComponentClass<ListProps, {}> = styled(List)`
  &&&&& .description {
    font-size: smaller;
  }
`;

type Props = {
  searchItems: any[];
};

let searchItem: any;
let index: number;
export const ProcessListView: React.SFC<Props> = ({ searchItems }) => (
  <React.Fragment>
    <Header icon="find" content="Search" subheader="Find process resources" />
    <Choose>
      <When condition={searchItems && searchItems.length > 0}>
        <VersionList divided relaxed>
          <For each="searchItem" of={searchItems} index="index">
            <ProcessListItemView key={index} searchItem={searchItem} />
          </For>
        </VersionList>
      </When>
      <Otherwise>
        <Message>There are no search results.</Message>
      </Otherwise>
    </Choose>
  </React.Fragment>
);
