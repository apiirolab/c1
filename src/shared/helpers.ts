interface Indexable<T> {
  [index: string]: T;
}

interface Group<T> {
  key: string;
  values: T[];
}

export function groupBy<T>(xs: T[], key: string): Indexable<T[]> {
  return xs.reduce(function(previous: any, current: Indexable<any>) {
    // Indexable<Array<Indexable<T>>>
    (previous[current[key]] = previous[current[key]] || []).push(current);
    return previous;
  }, {});
}

export function groupByArray<T>(xs: T[], key: string | Function): Array<Group<T>> {
  return xs.reduce(function(previous, current: Indexable<any>) {
    let v = key instanceof Function ? key(current) : current[key];
    let el = previous.find(r => r && r.key === v);
    if (el) {
      el.values.push(current);
    } else {
      previous.push({
        key: v,
        values: [current]
      });
    }
    return previous;
  }, []);
}
