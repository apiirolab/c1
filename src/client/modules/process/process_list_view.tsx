import * as React from 'react';
import { Query, QueryResult } from 'react-apollo';

import { ProcessListView } from '../views/process_list_view';
import { ProcessListQuery } from './process_queries';

function renderResult<T extends QueryResult<any, any>>(
  result: T,
  component: JSX.Element
): JSX.Element {
  if (result.loading) {
    return <p>Loading...</p>;
  }
  if (result.error) {
    return <p>Error :(</p>;
  }
  return component;
}

export const ProcessListContainer = () => (
  <ProcessListQuery>
    {result => renderResult(result, <ProcessListView searchItems={result.data.processes} />)}
  </ProcessListQuery>
);
