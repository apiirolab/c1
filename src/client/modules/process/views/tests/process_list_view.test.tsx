import * as React from 'react';

import { create as render } from 'react-test-renderer';

import { createData } from '../../../../../shared/test_data';
import { ProcessListView } from '../process_list_view';

describe('Process', () => {
  describe('List', () => {
    storyOf(
      'Empty',
      {
        componentWithData(searchItems: Corpix.Entities.Search[]) {
          return (
            <div>
              <ProcessListView searchItems={searchItems} />
            </div>
          );
        }
      },
      data => {
        it('renders no processes', () => {
          const root = render(data.componentWithData([]));
          expect(root).toMatchSnapshot();
        });
      }
    );

    storyOf(
      'With Data',
      {
        searchItems: [
          createData.searchItem({
            category: 'Travel Request Form',
            owner: 'Tomas Trescak',
            title: 'International Travel Request'
          }),
          createData.searchItem({
            category: 'Equipment Request Form',
            owner: 'Nguyen Vu',
            title: 'International Travel Request'
          }),
          createData.searchItem({
            category: 'Payment Request Form',
            owner: 'Tomas Trescak',
            title: 'International Travel Request'
          })
        ],
        componentWithData(searchItems: Corpix.Entities.Search[]) {
          searchItems = searchItems || this.searchItems;

          return (
            <div>
              <ProcessListView searchItems={searchItems} />
            </div>
          );
        }
      },
      data => {
        it('renders searched processes', () => {
          const root = render(data.componentWithData(data.searchItems));
          expect(root).toMatchSnapshot();
        });
      }
    );
  });
});
