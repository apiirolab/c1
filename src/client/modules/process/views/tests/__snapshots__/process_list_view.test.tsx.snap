// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process List Empty renders no processes 1`] = `
<div>
  <div
    className="ui header"
  >
    <i
      aria-hidden="true"
      className="find icon"
    />
    <div
      className="content"
    >
      Search
      <div
        className="sub header"
      >
        Find process resources
      </div>
    </div>
  </div>
  <div
    className="ui message"
  >
    There are no search results.
  </div>
</div>
`;

exports[`Process List With Data renders searched processes 1`] = `
<div>
  <div
    className="ui header"
  >
    <i
      aria-hidden="true"
      className="find icon"
    />
    <div
      className="content"
    >
      Search
      <div
        className="sub header"
      >
        Find process resources
      </div>
    </div>
  </div>
  <div
    className="ui divided relaxed list sc-bdVaJa bviDoP"
    role="list"
  >
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="file outline icon middle aligned"
      />
      <div
        className="content"
      >
        <a
          className="header"
        >
          Travel Request Form
        </a>
        <a
          className="description"
        >
          Submitted by 
          Tomas Trescak
           · 
           · 
          Thu, 01 Feb 2018 13:00:00 GMT
        </a>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="file outline icon middle aligned"
      />
      <div
        className="content"
      >
        <a
          className="header"
        >
          Equipment Request Form
        </a>
        <a
          className="description"
        >
          Submitted by 
          Nguyen Vu
           · 
           · 
          Thu, 01 Feb 2018 13:00:00 GMT
        </a>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
      role="listitem"
    >
      <i
        aria-hidden="true"
        className="file outline icon middle aligned"
      />
      <div
        className="content"
      >
        <a
          className="header"
        >
          Payment Request Form
        </a>
        <a
          className="description"
        >
          Submitted by 
          Tomas Trescak
           · 
           · 
          Thu, 01 Feb 2018 13:00:00 GMT
        </a>
      </div>
    </div>
  </div>
</div>
`;
